js/private/yaml.bzl

```diff
@@ -0,0 +1,413 @@
+            if type(curr_result) == "dict":
+                curr_result = curr_result.setdefault(state["key"], _empty_value_for_state(kns_states[i + 1]))
+            else:
+                if state["index"] >= len(curr_result):
+                    curr_result.append(_empty_value_for_state(kns_states[i + 1]))
+                curr_result = curr_result[state["index"]]
+
+def _set_result_value(starlark, stack, value):
+    "Add a new value to the starlark result corresponding to the last pseudostate in the stack"
+    kns_states = _get_key_and_sequence_states(stack)
+    if len(kns_states) == 0:
+        starlark["result"] = value
+    else:
+        curr_result = starlark["result"]
+        for state in kns_states[0:-1]:
+            if type(curr_result) == "dict":
+                curr_result = curr_result[state["key"]]
+            else:
+                curr_result = curr_result[state["index"]]
+        if type(curr_result) == "dict":
+            curr_result[kns_states[-1]["key"]] = value
+        else:
+            curr_result.append(value)
+
+def _empty_value_for_state(state):
+    if state["id"] == _STATE.KEY:
+        return {}
+    elif state["id"] == _STATE.SEQUENCE:
+        return []
+    else:
+        fail("State %s has no empty type" % state["id"])
+
+def _peek(stack):
+    return stack[-1]
+
+def _parse_scalar(value):
+    value = value.strip()
+    if _is_int(value):
+        return int(value)
+    elif _is_float(value):
+        return float(value)
+    elif _is_bool(value):
+        return _to_bool(value)
+    elif value.startswith("'"):
+        return value.strip("'")
+    else:
+        return value.strip("\"")
+
+def _is_float(value):
+    return value.replace(".", "", 1).isdigit()
+
+def _is_int(value):
+    return value.isdigit()
+
+def _is_bool(value):
+    return value == "true" or value == "false"
+
+def _to_bool(value):
+    if value == "true":
+        return True
+    elif value == "false":
+        return False
+    fail("Cannot convert scalar %s to a starlark boolean" % value)
+
+def _parse_key(key):
+    if key.startswith("'"):
+        return key.strip("'")
+    return key.strip("\"")
+
+def _get_key_and_sequence_states(stack):
+    return [state for state in stack if state["id"] in [_STATE.KEY, _STATE.SEQUENCE]]
+
+def _in_flow(stack):
+    kns_states = _get_key_and_sequence_states(stack)
+    return len(kns_states) > 0 and kns_states[-1]["id"] in [_STATE.SEQUENCE, _STATE.KEY] and kns_states[-1]["flow"]
+
+def _in_flow_sequence(stack):
+    kns_states = _get_key_and_sequence_states(stack)
+    return len(kns_states) > 0 and kns_states[-1]["id"] == _STATE.SEQUENCE and kns_states[-1]["flow"]
+
+def _in_flow_map(stack):
+    kns_states = _get_key_and_sequence_states(stack)
+    return len(kns_states) > 0 and kns_states[-1]["id"] == _STATE.KEY and kns_states[-1]["flow"]
+
+def _pop_higher_indented_states(stack, indent):
+    remove = []
+    for state in stack:
```

Comment:
nit: seems like instead of two iterations to collect removals and then apply, you could do one iteration where you construct a list to keep and return that?

Reply:
So the issue here is that I wanted to do an in-place removal, and starlark complains if you try to remove an element inside of a loop.